iron-form.html

```diff
@@ -11,503 +11,459 @@
+        this._form.addEventListener('reset', this.reset.bind(this));
+
+        // Save the initial values.
+        this._defaults = this._defaults || new WeakMap();
+        var nodes = this._getSubmittableElements();
+        for (var i = 0; i < nodes.length; i++) {
+          var node = nodes[i];
+          if (!this._defaults.has(node)) {
+            this._defaults.set(node, {
+              checked: node.checked,
+              value: node.value,
+            });
+          }
+        }
+      },
 
-      // Allow for a presubmit hook
-      var event = this.fire('iron-form-presubmit', {}, {cancelable: true});
-      if(!event.defaultPrevented) {
-        this.request.generateRequest();
-        this.fire('iron-form-submit', json);
-      }
-    },
-
-    /**
-     * Handler that is called when the native form fires a `submit` event
-     *
-     * @param {Event} event A `submit` event.
-     */
-    _onSubmit: function(event) {
-      this.submit();
-
-      // Don't perform a page refresh.
-      if (event) {
-        event.preventDefault();
-      }
+      /**
+       * Validates all the required elements (custom and native) in the form.
+       * @return {boolean} True if all the elements are valid.
+       */
+      validate: function() {
+        if (this._form.getAttribute('novalidate') === '')
+          return true;
 
-      return false;
-    },
-
-    /**
-     * Handler that is called when the native form fires a `reset` event
-     *
-     * @param {Event} event A `reset` event.
-     */
-    _onReset: function(event) {
-      this._resetCustomElements();
-    },
-
-    /**
-     * Returns a json object containing name/value pairs for all the registered
-     * custom components and native elements of the form. If there are elements
-     * with duplicate names, then their values will get aggregated into an
-     * array of values.
-     *
-     * @return {!Object}
-     */
-    serialize: function() {
-      var json = {};
-
-      function addSerializedElement(name, value) {
-        // If the name doesn't exist, add it. Otherwise, serialize it to
-        // an array,
-        if (json[name] === undefined) {
-          json[name] = value;
-        } else {
-          if (!Array.isArray(json[name])) {
-            json[name] = [json[name]];
+        // Start by making the form check the native elements it knows about.
+        var valid = this._form.checkValidity();
+        var elements = this._getValidatableElements();
+
+        // Go through all the elements, and validate the custom ones.
+        for (var el, i = 0; el = elements[i], i < elements.length; i++) {
+          // This is weird to appease the compiler. We assume the custom element
+          // has a validate() method, otherwise we can't check it.
+          var validatable = /** @type {{validate: (function() : boolean)}} */ (el);
+          if (validatable.validate) {
+            valid = !!validatable.validate() && valid;
```

Comment:
Oh this doesn't work: you still need to validate _all_ the elements in the form (because this triggers invalid UI and stuff)